Varnish FieldProps Implementation

bska opened the pull request:
In particular,

  * Prefer in-place map element construction where possible, and `insert_or_assign()` where not
  * Apply `const` where meaningful
  * Hoist constant conditions/objects out of loops
  * Split long lines and make a few other whitespace adjustments
  * Prefer prefix increment and `//` comments
  * Remove an unneeded static object
  * Add `{}` to single statement control structures
  * Separate `OPERATER` handling out from the other `*REG` operations (`ADDREG`, `EQUALREG`, `MULTIREG`).  The distinctions are greater than the commonalities and that justifies separate handling.

This is in preparation of making our implementation of array operations like `COPY` more strict.

bska commented:
I am creating this PR in draft mode because it depends on, and contains, the earlier PR #4102.  I will keep the PR in a draft state until such time as it is ready for review and merging.

bska commented:
jenkins build this please

bska commented:
jenkins build this please

bska commented:
The prerequisite PR #4102 was merged into master.  I'm marking this PR as "ready for review" now and I'm running a build check.

bska commented:
jenkins build this please

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldData.hpp:
```diff
@@ -22,92 +22,110 @@
+#include <string>
+#include <vector>
+
+namespace Opm { namespace Fieldprops {
```
since there only are Opm::FieldProps members this can be namespace Opm::Fieldprops

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -206,50 +207,54 @@ global_kw_info(const std::string& name, bool)
+//
+// This quite weird behaviour comes from reading the GRIDOPTS and MULTNUM
+// documentation, and practical experience with ECLIPSE simulations.
+// Ufortunately the documentation of the xxxREG keywords does not confirm
```
Unfortunately

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -691,144 +775,185 @@ void FieldProps::distribute_toplayer(Fieldprops::FieldData<double>& field_data,
-        multiplier_kw_infos_[mult_keyword] = kw_info;
+    }
+    else if (multiplier_in_edit) {
+        assert((keyword != ParserKeywords::PORV::keywordName) &&
```
general niggle. asserting with && makes assertion less useful, multiple statements is better..

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -887,256 +1012,389 @@ void FieldProps::apply_multipliers()
-  therefor not make sense to require that these fields are fully defined.
-*/
+// The ACTNUM and PORV keywords are special cased with quite extensive
+// postprocessing, and should be access through the special ::porv() and
```
accessed

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -887,256 +1012,389 @@ void FieldProps::apply_multipliers()
+// keywords are also hidden from the keys<T>() vectors.
+//
+// If there are TRAN? fields in the container they are transferred even if they
+// are not completely defined. This is because that TRAN? fields will ultimately
```
-that

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -1364,25 +1685,24 @@ const std::string& FieldProps::canonical_fipreg_name(const std::string& fipreg)
-  FieldProps instance.
-*/
-std::vector<int> FieldProps::actnum() {
+// This function generates a new ACTNUM vector.The ACTNUM vector which is
```
Space after dot

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -1364,25 +1685,24 @@ const std::string& FieldProps::canonical_fipreg_name(const std::string& fipreg)
-*/
-std::vector<int> FieldProps::actnum() {
+// This function generates a new ACTNUM vector.The ACTNUM vector which is
+// returned is joined result of three different data sources:
```
Suggested reformulation: The returned ACTNUM vector is a joined result from three different data sources:

akva2 commented on opm/input/eclipse/EclipseState/Grid/FieldProps.cpp:
```diff
@@ -1364,25 +1685,24 @@ const std::string& FieldProps::canonical_fipreg_name(const std::string& fipreg)
+// This function generates a new ACTNUM vector.The ACTNUM vector which is
+// returned is joined result of three different data sources:
+//
+//    1. The ACTNUM of if the grid which is part of this FieldProps structure.
```
-if

akva2 reviewed (approved):
Some suggestions for further polish.

bska commented:
Thank you for the review.  I have pushed an update that addresses your comments.  Please let me know if I've missed some.

bska commented:
>> I have pushed an update that addresses your comments. Please let me know if I've missed some.
>
> looks all good now.

Much appreciated.  I'll run a final build check just to be sure that I have not accidentally introduced a problem in the latest update.

bska commented:
jenkins build this please

bska commented:
Build check is green and the PR is approved.  I'll merge into master.
